namesys/namesys.go

```diff
@@ -1,59 +1,95 @@
 package namesys
 
 import (
+	"strings"
+
 	context "github.com/ipfs/go-ipfs/Godeps/_workspace/src/golang.org/x/net/context"
 	ci "github.com/ipfs/go-ipfs/p2p/crypto"
 	path "github.com/ipfs/go-ipfs/path"
 	routing "github.com/ipfs/go-ipfs/routing"
 )
 
-// ipnsNameSystem implements IPNS naming.
+// ipfs NameSystem implements generic IPFS naming.
```

Comment:
On Sat, May 09, 2015 at 09:57:01AM -0700, Juan Batiz-Benet wrote:
> > -// ipnsNameSystem implements IPNS naming.
> > +// ipfs NameSystem implements generic IPFS naming.
> …
> this is not about godoc or the ipfs implementations-- it's about how
> we describe the system that resolves:
> 
> ```
> /ipns/Qm..../...
> /ipns/foo.com/...

I think that's also about the godoc.  We should have a canonical
explanation for what we're about here, and the godoc is a good place
to store that.  If we want a more elaborate form in the docs (maybe in
the tour?) that's fine, but I think we need to clarify what goes into
the godocs before biting that off.

And as I've said earlier [1], I want a separate directory for
/dns/foo.com/… to clearly distinguish between DNS-based links and
DHT-based links.  I think using “IPNS” for both DNS-links and
DHT-links is going to cause more confusion than having a single
directory saves you.  So far, I've been suggesting “IPNS” just apply
to the DHT links, but I'd also be ok with:

  /ipns/dht/Qm…
  /ipns/dns/foo.com…

So what I want:

1. Explicit prefixes/namespaces for the different protocols.  I expect
   this will make the system *less* confusing, by avoiding ambiguity
   about which protocol is resolving which name.  That helps:
   a. Users, by making it easy to identify (and find docs, file bugs,
      …) for the protocol they're using.
   b. Protocol authors, who only need to register a unique
      prefix/namespace for their protocol, and not worry about making
      sure their hex-encoded frobnitz references aren't accidentally
      treated as base-58 multihashes.
   c. Sysadmins, who don't need to worry about resolution order when
      they're configuring the resolver to decide which protocol gets
      to interpret an ambiguous name.

2. An explicit name for SFS-style references stored in the DHT.  I've
   been calling this system “IPNS”, but if that's supposed to be
   mutable links in general, we'll need a different name here (and
   “SFS-style PKI names” isn't an easy-to-search, unambiguous name ;).

3. A clear procedure for protocol authors to integrate their mutable
   name system with the IPFS-world tooling.  This would be trivial
   with FUSE, just return symlinks:

     /frobnitz/frobby-key → /dns/foo.com
     /frobnitz/nitzy-key → /ipfs/QmSomeKey/a/b/c

   For getting into ‘ipfs resolve …’ we'll probably want something
   that the resolver can launch as a subprocess or call as a service
   if it doesn't have a built-in resolver for a given protocol name.

> … it's less pure than inventing another abstraction, but the
> description:
> 
> > "ipns supports sfs-style pki based names, and other resolvers (dns
> > + ...)."
> 
> is pure enough and gives end users less to learn.

I don't think many users will care one way or the other if the mutable
name system is:

  /ipns/<all-the-keys>

or:

  /ipns/<protocol>/<protocol-specific-keys>

or:

  /<protocol>/<protocol-specific-keys>

Personally, I favor the last form to mirror
/ipfs/<ipfs-specific-keys>.  Do you really think it's a big usability
win to drop the protocol namespace?  It makes the implementation so
much trickier…